### PR TITLE
{lcrq,librecast,lcsync}: init at {0.1.0, 0.7-RC3, 0.2.1}

### PR DESCRIPTION
**Pull request by jleightcap**
## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/sync/lcsync/default.nix
Comment (drupol) [inline pkgs/applications/networking/sync/lcsync/default.nix:8]: Remove this

File: pkgs/applications/networking/sync/lcsync/default.nix
Comment (drupol) [inline pkgs/applications/networking/sync/lcsync/default.nix:10]: I think we should use the `finalAttrs` pattern here.

The `finalAttrs` pattern will let you remove the `rec` keyword ([see its implementation in Nix](https://github.com/NixOS/nix/blob/2d1d81114d72ace89ce08cd3bc93f4eb27a2975d/src/libexpr/nixexpr.cc#L121)).

Why is this pattern preferred to `rec` ?

Let's take this simple code example:

```nix
mkDerivation rec {
   foo = 1;
   bar = foo + 1;
}
```

and then `.overrideAttrs(old: { foo = 2; })`, you'll get `{ foo = 2; bar = 2; }` while with `finalAttrs` pattern, it would work correctly because it's a real fixed point.

Let me share a couple of useful links regarding the `finalAttrs` pattern:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

Feel free to reach out if you need some assistance.

File: pkgs/applications/networking/sync/lcsync/default.nix
Comment (drupol) [inline pkgs/applications/networking/sync/lcsync/default.nix:19]: Prefer `hash = ...` to `sha256 = ...` Both currently work, but the former will make it easier if the algorithm ever needs to change.

Since there could be different hashing algorithms in use, this make sense.

File: pkgs/applications/networking/sync/lcsync/default.nix
Comment (drupol) [inline pkgs/applications/networking/sync/lcsync/default.nix:3]: Never pass `pkgs`, just pass the strict necessary dependencies.

File: pkgs/applications/networking/sync/lcsync/default.nix
Comment (drupol) [inline pkgs/applications/networking/sync/lcsync/default.nix:21]:
```suggestion
  buildInputs = [ lcrq librecast libsodium ];
```

File: pkgs/applications/networking/sync/lcsync/default.nix
Comment (drupol) [inline pkgs/applications/networking/sync/lcsync/default.nix:26]: Keep `meta` attributes sorted alphabetically.

**Review by drupol — CHANGES_REQUESTED** [review]
Hi,

Thank you once again for your valuable contribution.

However, there is an issue in the way you're attempting to introduce those software packages into `nixpkgs`. The 3 PRs you've submitted (https://github.com/NixOS/nixpkgs/pull/248430, https://github.com/NixOS/nixpkgs/pull/248431, https://github.com/NixOS/nixpkgs/pull/248437) need to be thoroughly reevaluated.

Creating a separate PR for each package is problematic, as it leads to code duplication. For instance, in the process of packaging `lcsync`, you've included `lcrq`, even though there's already a PR for `lcrq`.

We kindly request you to clean up this redundancy, and you can do this in one of two ways:

1. **Single PR Method:** Create one consolidated PR with 3 commits, containing `librecast`, `lcrq`, and `lcsync`.
2. **Multiple PR Method:** If you prefer multiple PRs, each containing only one single software, then please ensure that you wait for the merge of one before submitting the next. This will help to avoid duplicating code.

Thank you for your cooperation!

**Comment by albertchae** [comment]
Hi @drupol, thanks for the thorough feedback on all 3 PRs and apologies for the confusion. We didn't intend for them to be reviewed until https://github.com/NixOS/nixpkgs/pull/248434 was merged, which is why we left them in draft state and was planning on flipping them to reviewable one by one as they got merged. In the future, we will avoid opening PRs at all even in draft state to minimize reviewer burden!

We addressed all the comments across the 3 PRs and consolidated them here. Hopefully we didn't miss any but let us know if there's anything else we need to fix 🙏

File: pkgs/development/libraries/lcrq/default.nix
Comment (jasonodoom) [inline pkgs/development/libraries/lcrq/default.nix:26]: We originally had `platforms.unix` but we ran into a compilation issue with `gcc` vs `clang` on darwin. Before this change we receive the below error:

```
> ld: unknown option: -soname
```

According to [this issue](https://github.com/CopernicaMarketingSoftware/PHP-CPP/issues/368#issuecomment-367825982) this is a known difference between clang+llvm and gcc. Is there a workaround or is it more appropriate to report to the project authors?

Comment (drupol) [reply pkgs/development/libraries/lcrq/default.nix:26]: Dear @NixOS/darwin-maintainers ,

Can you help with this comment?

Thanks

Comment (emilazy) [reply pkgs/development/libraries/lcrq/default.nix:26]: This should definitely go upstream, but maybe try

```nix
postPatch = lib.optionalString stdenv.isDarwin ''
  substituteInPlace src/Makefile.in \
    --replace .so .dylib \
    --replace -soname -install_name
'';
```

**Review by drupol — APPROVED** [review]

**Comment by drupol** [comment]
Congrats, nice addition !

**Comment by onepict** [comment]
From: https://todo.sr.ht/~librecast/lcrq/8

"Darwin lacks htobe32 & htobe64, among the other functions in that family for bite swapping. 
Try using a [wrapper](https://gist.github.com/panzi/6856583#file-portable_endian-h-L20) which makes use of the functions in [libkern/OSByteOrder.h](https://github.com/apple-oss-distributions/xnu/blob/xnu-8796.141.3/libkern/libkern/OSByteOrder.h)"

Brett Sheffield also put a call out as we have no access to test this architecture.  

https://chaos.social/@dentangle/110909926788632013

**Pull request merged**
